README.md
# node-rtsp-stream-es6

First of all, it's a fork of a [fork](https://github.com/Wifsimster/node-rtsp-stream-es6) of [**node-rtsp-stream**](https://www.npmjs.com/package/node-rtsp-stream)

## Differences with the original modules

- Stream can be started and stopped from the browser to save resources on the server.
- Written in ES6 instead of CoffeeScript
- Github repository

## Description

Stream any RTSP stream and output to [WebSocket](https://github.com/websockets/ws) for consumption by [jsmpeg](https://github.com/phoboslab/jsmpeg).
HTML5 streaming video!

**NOTE:** This fork will automatically start/stop the FFMPEG stream on the backend server when the first client connects and the last client disconnects.  There is a delay between the last client disconnecting and the `SIGINT` being sent which can be customized in the options. If a new client reconnects within the timeout, the stream continues to run.

## Requirements

You need to download and install [FFMPEG](https://ffmpeg.org/download.html).

## Installation

```shell
npm i shbatm/node-rtsp-stream-es6
```

## Server

```javascript
Stream = require('node-rtsp-stream-es6');

const options = {
    name: 'streamName',
    url: 'rtsp://184.72.239.149/vod/mp4:BigBuckBunny_115k.mov',
    ffmpegPort: 3000,
    protocol: "tcp", // Optional "tcp" or "udp", default: "tcp"
    frameRate: "30", // Optional - use to set the frame rate of the stream
    shutdownDelay: 10000,  // Optional - delay before FFMPEG stopping after last client disconnects
    hideFfmpegOutput: false, // Optional - show/hide FFMPEG console output
                            // As of v1.1.2 - default is true.
    enableAudio: false, // Default is false. Set to true to enable audio in FFMPEG
    hwAccel: false, // Attempt to use hardware accelerated codecs
    ffmpegCustomArgs: [] // Array of custom FFMPEG CLI arguments. WARNING: this will overrite all ffmpeg options and EVERYTHING must be passed as an array of strings (including stream url, protocol, and framerate). Only use if you know what you are doing!
};

stream = new Stream(options);
stream.start();
```

## Client

You can find a live stream JSMPEG example here : https://github.com/phoboslab/jsmpeg/blob/master/view-stream.html


CHANGELOG.md
### [1.1.7] Incorporate upstream changes, refactor, add harware acceleration

* Changed function names and initialization to match upstream fork
* Passed `options` object directly to mpeg1muxer instead of reassigning it several times
* Added `hwAccel` option to attempt to use hardware accelerated version of FFMPEG.

### [1.1.4] Change port parameter to ffmpegPort

* Change name of port parameter for clarification purposes.

### [1.1.3] Disable Audio and Add Custom FFMPEG Options

* Audio is disabled by default. Pass `enableAudio: true` to enable.
* To pass a custom array of FFMPEG arguements, pass `ffmpegCustomArgs` in the options.

### [1.1.2] Bug fixes

* Bug fixes from v1.1.1
* `hideFfmpegOutput` now defaults to `false` if not passed as an option

### [1.1.1] Minor update

* Added TCP/UDP Protocol Option

### [1.1.0] Initial Forked Release

* Initial release of original forked version


test/index.js
/* jshint node: true, esversion: 6*/
const app = require('express')();
const http = require('http').Server(app);
const ioc = require('socket.io-client');
const port = 3000;
const socket = ioc('http://localhost:' + port);

app.get('/', (req, res) => {
  res.send('<h1>Hello world</h1>');
});

http.listen(80, () => {
  console.log('Listening on *:80');
});

socket.on('connect', () => { console.log('Connected !'); });

socket.on('stream', (data) => { console.log(data); });

test/client.js
/* jshint node: true, esversion: 6*/
const WebSocket = require('ws');
const ws = new WebSocket('ws://localhost:3000');

ws.on('open', () => {
  console.log('Connected to stream');
});

ws.on('message', (data, flags) => {
  console.log(data);
});

test/server.js
/* jshint node: true, esversion: 6*/
Stream = require('node-rtsp-stream-es6');

const options = {
    name: 'streamName',
    url: 'rtsp://184.72.239.149/vod/mp4:BigBuckBunny_115k.mov',
    ffmpegPort: 3000,
    frameRate: "30", // Optional - use to set the frame rate of the stream
    shutdownDelay: 10000,  // Optional - delay before FFMPEG stopping after last client disconnects
    hideFfmpegOutput: false // Optional - show/hide FFMPEG console output
};

stream = new Stream(options)

stream.start()

setTimeout(stream.stop.bind(stream), 10 * 1000)


src/mpeg1muxer.js
/* jshint node: true, esversion: 6*/

const child_process = require('child_process');
const EventEmitter = require('events');

class Mpeg1Muxer extends EventEmitter {

    constructor(options) {
        super(options);

        var ffmpegArgs = ["-rtsp_transport", options.protocol, "-r", options.frameRate, "-i", options.url, "-f", "mpegts", "-codec:v", "mpeg1video", "-bf", "0", "-"];

        if (options.enableAudio) {
            ffmpegArgs.splice(-3, 0, '-codec:a', 'mp2');
        } else {
            ffmpegArgs.splice(-3, 0, "-an");
        }
        if (options.hwAccel) {
            ffmpegArgs.splice(2, 0, "-codec:v", "h264_mmal");
            // ffmpegArgs.splice(ffmpegArgs.indexOf("mpeg1video"), 1, "h264_omx");
        }
        if (options.ffmpegCustomArgs) {
            ffmpegArgs = options.ffmpegCustomArgs;
        }

        if (!options.hideFfmpegOutput) {
            console.log("Launching ffmpeg " + ffmpegArgs.join(' '));
        }

        this.stream = child_process.spawn("ffmpeg", ffmpegArgs, {
            detached: false
        });

        this.inputStreamStarted = true;
        this.stream.stdout.on('data', (data) => {
            return this.emit('mpeg1data', data);
        });
        this.stream.stderr.on('data', (data) => {
            return this.emit('ffmpegError', data);
        });
    }

    stop() {
        this.stream.stdout.removeAllListeners();
        this.stream.stderr.removeAllListeners();
        this.stream.kill();
        this.stream = undefined;
    }
}

module.exports = Mpeg1Muxer;

src/videoStream.js
/* jshint node: true, esversion: 6*/

const WebSocket = require('ws');
const EventEmitter = require('events');
const STREAM_MAGIC_BYTES = "jsmp";
const Mpeg1Muxer = require('./mpeg1muxer');
class VideoStream extends EventEmitter {

    constructor(options) {
        super(options);

        // Set defaults if not provided
        options.protocol = options.protocol || "tcp";
        options.frameRate = options.frameRate || "30";
        options.shutdownDelay = options.shutdownDelay || 5000;
        options.hideFfmpegOutput = ("hideFfmpegOutput" in options) ? options.hideFfmpegOutput : true;
        options.hwAccel = ("hwAccel" in options) ? options.hwAccel : false;

        if (!options.hideFfmpegOutput) { console.log(JSON.stringify(options, undefined, 3)); }

        // Finish setup
        this.stream = void 0;
        this.connectionCount = 0;
        this.streamActive = false;
        this.disconnectTimeout = null;
        this.options = options;
        this.stream2Socket();
    }

    stream2Socket() {
        console.log(`Starting WebSocket server on port ${this.options.ffmpegPort}. Waiting for connections...`);
        this.server = new WebSocket.Server({ port: this.options.ffmpegPort });
        this.server.on('connection', (socket) => {
            if (this.connectionCount === 0) {
                if (!this.disconnectTimeout) {
                    // First connection, start the stream.
                    this.start();
                    this.streamActive = true;
                } else {
                    // New connection while still in shutdown timeout.
                    clearTimeout(this.disconnectTimeout);
                    this.disconnectTimeout = null;
                    this.streamActive = true;
                }
            }
            this.connectionCount++;
            console.log(`${this.options.name}: new connection! ${this.connectionCount} active connections.`);

            let streamHeader = new Buffer(8);
            streamHeader.write(STREAM_MAGIC_BYTES);
            streamHeader.writeUInt16BE(this.options.width, 4);
            streamHeader.writeUInt16BE(this.options.height, 6);
            socket.send(streamHeader);

            socket.on('close', () => {
                this.connectionCount--;
                console.log(`${this.options.name} client disconnected! ${this.connectionCount} active connections.`);
                if (this.connectionCount === 0) {
                    this.stopStream(this.shutdownDelay);
                }
            });
        });

        this.on('camdata', (data) => {
            this.server.clients.forEach((client) => {
                if (client.readyState === WebSocket.OPEN) { client.send(data); }
            });
        });
    }

    start() {
        this.mpeg1Muxer = new Mpeg1Muxer(this.options);
        this.mpeg1Muxer.on('mpeg1data', (data) => {
            return this.emit('camdata', data);
        });

        let gettingInputData = false;
        let gettingOutputData = false;
        let inputData = [];
        let outputData = [];

        this.mpeg1Muxer.on('ffmpegError', (data) => {
            data = data.toString();
            if (data.indexOf('Input #') !== -1) { gettingInputData = true; }
            if (data.indexOf('Output #') !== -1) {
                gettingInputData = false;
                gettingOutputData = true;
            }
            if (data.indexOf('frame') === 0) { gettingOutputData = false; }
            if (gettingInputData) {
                inputData.push(data.toString());
                let size = data.match(/\d+x\d+/);
                if (size !== null) {
                    size = size[0].split('x');
                    if (this.options.width === null) { this.options.width = parseInt(size[0], 10); }
                    if (this.options.height === null) { this.options.height = parseInt(size[1], 10); }
                }
            }
        });
        if (!this.options.hideFfmpegOutput) {
            this.mpeg1Muxer.on('ffmpegError', (data) => {
                return global.process.stderr.write(data);
            });
        }
        return this;
    }

    stop(serverCloseCallback) {
        this.server.close(serverCloseCallback);
        this.server.removeAllListeners();
        this.server = undefined;

        this.mpeg1Muxer.stop();
        this.mpeg1Muxer.removeAllListeners();
        this.mpeg1Muxer = undefined;
    }

    stopStream(timeout) {
        console.log(`Last connection closed! Stopping stream in ${timeout/1000} seconds...`);
        this.disconnectTimeout = setTimeout(() => {
            this.stop();
            this.disconnectTimeout = null;
            this.streamActive = false;
            console.log("Stream closed. Server still running waiting for new connections...");
        }, timeout);
    }
}

module.exports = VideoStream;
